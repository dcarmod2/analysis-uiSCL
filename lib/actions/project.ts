import {stringify} from 'querystring'
import {createAction} from 'redux-actions'

import {API} from 'lib/constants'
import selectCurrentProject from 'lib/selectors/current-project'

import fetch from './fetch'
import getFeedsRoutesAndStops from './get-feeds-routes-and-stops'
import {
  saveToServer as saveModification,
  getForProject as getModificationsForProject
} from './modifications'

import {loadBundle} from './'

const PROJECT_URL = API.Project

// project stuff
const addProject = createAction('add project')
export const create = ({bundleId, name, regionId}) =>
  fetch({
    url: PROJECT_URL,
    options: {
      body: {
        bundleId,
        name,
        regionId,
        variants: ['Default']
      },
      method: 'post'
    },
    next: (response) => addProject(response.value)
  })

const deleteLocally = createAction('delete project')
export const deleteProject = (projectId) =>
  fetch({
    options: {
      method: 'delete'
    },
    url: `${PROJECT_URL}/${projectId}`,
    next: () => deleteLocally(projectId)
  })

export const loadProject = (_id) =>
  fetch({
    url: `${PROJECT_URL}/${_id}`,
    next: (r) => set(r.value)
  })

export const loadProjectAndModifications = (_id) => async (dispatch) => {
  const r1 = await Promise.all([
    dispatch(loadProject(_id)),
    dispatch(getModificationsForProject(_id))
  ])
  const project = r1[0]
  const modifications = r1[1]

  const r2 = await Promise.all([
    dispatch(loadBundle(project.bundleId)),
    dispatch(
      getFeedsRoutesAndStops({
        bundleId: project.bundleId,
        forceCompleteUpdate: true,
        modifications
      })
    )
  ])

  return {bundle: r2[0], feeds: r2[1], modifications, project}
}

export const saveToServer = (project) =>
  fetch({
    url: `${PROJECT_URL}/${project._id}`,
    options: {
      body: project,
      method: 'put'
    },
    next: (response) => set(response.value)
  })
export const set = createAction('set project')
export const setAll = createAction('set projects')
export const loadProjects = (query = {}) =>
  fetch({
    url: `${PROJECT_URL}?${stringify(query)}`,
    next: (response) => setAll(response.value)
  })

/**
 * Project Variants
 */
export const createVariant = (name) => (dispatch, getState) => {
  const project = selectCurrentProject(getState())
  const {variants} = project
  return dispatch(
    saveToServer({
      ...project,
      variants: [...variants, name]
    })
  )
}

/**
 * Create a new scenario with all of the same modifications,
 */
export const copyScenario = (index) => async (dispatch, getState) => {
  const state = getState()
  const project = selectCurrentProject(state)
  const {modifications} = state.project

  const newScenarioName = project.variants[index] + ' (copy)'
  const oldLength = project.variants.length

  await dispatch(
    saveToServer({
      ...project,
      variants: [...project.variants, newScenarioName]
    })
  )

  return Promise.all(
    modifications.map((m) => {
      const variants = m.variants
      // Could come from a bitset on the Java side so may be of varying length
      for (let i = 0; i < oldLength; i++) {
        if (variants[i] === undefined) variants[i] = false
      }

      return dispatch(
        saveModification({
          ...m,
          variants: [
            ...variants,
            variants[index] // add the new variant
          ]
        })
      )
    })
  )
}

export const deleteVariant = (index) => async (dispatch, getState) => {
  const state = getState()
  const project = selectCurrentProject(state)
  const {modifications} = state.project

  await dispatch(
    saveToServer({
      ...project,
      variants: project.variants.filter((_, i) => i !== index)
    })
  )

  return Promise.all(
    modifications.map((m) =>
      dispatch(
        saveModification({
          ...m,
          variants: m.variants.filter((_, i) => i !== index)
        })
      )
    )
  )
}

export const editVariantName = ({index, name}) => (dispatch, getState) => {
  const project = selectCurrentProject(getState())
  return dispatch(
    saveToServer({
      ...project,
      variants: project.variants.map((value, i) => (i === index ? name : value))
    })
  )
}
