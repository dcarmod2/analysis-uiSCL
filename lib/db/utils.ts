import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
