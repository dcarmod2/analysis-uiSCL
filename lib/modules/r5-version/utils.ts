import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
