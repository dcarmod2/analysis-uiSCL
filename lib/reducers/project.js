import sortBy from 'lodash/sortBy'

function createInitialState() {
  return {
    modifications: [],
    feeds: [],
    projects: []
  }
}

const sortByTypeName = (a) => sortBy(a, ['type', 'name'])
const sortByCreatedAt = (a) => sortBy(a, (i) => -i.createdAt)

export const reducers = {
  /** add a newly created project to the state */
  'add project'(state, action) {
    return {
      ...state,
      projects: [action.payload, ...state.projects]
    }
  },
  'delete modification'(state, action) {
    const modificationId = action.payload
    return {
      ...state,
      modifications: state.modifications.filter((m) => m._id !== modificationId)
    }
  },
  'delete project'(state, action) {
    return {
      ...state,
      projects: state.projects.filter((p) => p._id !== action.payload)
    }
  },
  'log out'() {
    return createInitialState()
  },
  'set feeds'(state, action) {
    return {
      ...state,
      feeds: action.payload || []
    }
  },
  'create modification'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName([action.payload, ...state.modifications])
    }
  },
  'create modifications'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName([...state.modifications, ...action.payload])
    }
  },
  'set modification'(state, action) {
    const modification = action.payload
    const modifications = state.modifications.filter(
      (m) => m._id !== modification._id
    )
    return {
      ...state,
      modifications: sortByTypeName([...modifications, modification])
    }
  },
  'set modifications'(state, action) {
    return {
      ...state,
      modifications: sortByTypeName(action.payload)
    }
  },
  'update modification'(state, action) {
    const {_id, ...newProps} = action.payload
    return {
      ...state,
      modifications: state.modifications.map((m) =>
        m._id === _id ? {...m, ...newProps} : m
      )
    }
  },
  'set project'(state, action) {
    const {projects} = state
    const currentProject = {...action.payload}
    return {
      ...state,
      projects: sortByCreatedAt([
        ...projects.filter((p) => p._id !== currentProject._id),
        currentProject
      ])
    }
  },
  'set projects'(state, action) {
    return {
      ...state,
      projects: action.payload
    }
  },
  'clear current region'() {
    return createInitialState()
  }
}

export const initialState = createInitialState()
