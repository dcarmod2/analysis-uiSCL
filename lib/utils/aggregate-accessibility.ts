import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
