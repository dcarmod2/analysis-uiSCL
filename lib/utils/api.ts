// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
