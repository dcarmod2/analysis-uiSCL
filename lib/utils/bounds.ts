import lonlat from '@conveyal/lonlat'

import L from 'lib/leaflet'
import message from 'lib/message'

import reprojectCoordinates from './reproject-coordinates'

const ANTIMERIDIAN_WARNING =
  'We do not currently support bounds that cross the antimeridian.'

type Bounds = {
  north: number
  south: number
  east: number
  west: number
  toString: () => string
}

export function toLatLngBounds(
  bounds: Bounds | L.LatLngBounds
): L.LatLngBounds {
  // Check if it's already a leaflet bounds object
  if (bounds instanceof L.LatLngBounds) return bounds
  return new L.LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds(bounds: L.LatLngBounds) {
  // Don't change the original bounds object
  let b = new L.LatLngBounds(bounds.getSouthWest(), bounds.getNorthEast())

  if (b.getWest() < -180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), -180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), 180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds(bounds) {
  return new L.LatLngBounds(
    lonlat.toLeaflet(reprojectCoordinates(bounds.getSouthWest())),
    lonlat.toLeaflet(reprojectCoordinates(bounds.getNorthEast()))
  )
}
