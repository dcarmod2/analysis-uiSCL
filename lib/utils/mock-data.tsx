import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import React from 'react'
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'util'
import {constructor as XorShift} from 'xorshift'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN, TRAVEL_TIME_PERCENTILES} from '../constants'

import multi from './multi'

// Create a random number generator that is seeded
const seededRandom = new XorShift([1, 2, 3, 4])

export const makeMockStore = configureStore([multi, thunkMiddleware])

export const mockWithProvider = (children, mockData = mockStores.init) => {
  const store = makeMockStore(mockData)
  const wrapper = mount(<Provider store={store}>{children}</Provider>)

  return {
    snapshot: () => mountToJson(wrapper),
    store,
    wrapper
  }
}

export const mockModel = {
  accessGroup: 'accessGroupName',
  createdAt: '2017-11-08T12:30:50.133Z',
  createdBy: 'person@gmail.com',
  nonce: '12345',
  updatedAt: '2017-11-08T12:31:19.602Z',
  updatedBy: 'person@gmail.com'
}

export const mockSegmentGeom: GeoJSON.LineString = {
  type: 'LineString',
  coordinates: [
    [-122.0246, 36.9707],
    [-122.0279, 37.049],
    [-121.9799, 37.2299],
    [-121.9445, 37.324],
    [-121.936, 37.353],
    [-121.924, 37.365]
  ]
}

export const mockSegment: CL.ModificationSegment = {
  fromStopId: '1',
  toStopId: null,
  geometry: mockSegmentGeom,
  spacing: 3000,
  stopAtEnd: false,
  stopAtStart: false
}

export const mockStop1 = {
  autoCreated: false,
  bearing: 0,
  center: [36.9707, -122.0246],
  distanceFromStart: 0,
  index: 0,
  lat: 36.9707,
  lon: -122.0246,
  stopId: '1'
}

export const mockGtfsStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockGtfsStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockGtfsStops = [mockGtfsStop1, mockGtfsStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  name: 'Mock Pattern',
  pattern_id: 'mock-pattern-id',
  stops: mockGtfsStops,
  trips: [
    {
      duration: 1234,
      end_time: 13579,
      start_time: 12345,
      trip_headsign: 'To Downtown',
      trip_id: 'abcd',
      trip_short_name: 'The Express'
    }
  ]
}

export const mockRoute = {
  label: 'mock-route-label',
  patterns: [mockPattern],
  route_id: 'route1'
}

export const mockFeed = {
  checksum: 'MOCK',
  id: '1',
  routes: [mockRoute],
  stopsById: {
    1: mockGtfsStop1,
    2: mockGtfsStop2
  },
  stops: [mockGtfsStop1, mockGtfsStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  _id: 'timetable-id',
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  dwellTimes: [12],
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd',
  phaseFromTimetable: '',
  phaseFromStop: '',
  phaseAtStop: '',
  phaseSeconds: 300,
  exactTimes: false,
  modificationId: '1234',
  modificationName: 'Test Modification'
}

export const mockOpportunityDataset = {
  sourceName: 'lodes-data-2014',
  name: 'Workers total',
  _id: 'Workers_total'
}

const regionId = '1'

export const mockBundle = {
  ...mockModel,
  errorCode: '',
  name: 'Mock Bundle',
  _id: '1',
  centerLat: 38.90124,
  centerLon: -77.015615,
  feeds: [],
  regionId,
  serviceStart: '2018-09-27',
  serviceEnd: '2018-09-28',
  status: 'DONE',
  totalFeeds: 0
}

export const mockRegion = {
  ...mockModel,
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Region description',
  _id: regionId,
  name: 'Mock Region',
  opportunityDatasets: [mockOpportunityDataset],
  statusCode: 'DONE'
}

export const mockProject = {
  ...mockModel,
  bundleId: '1',
  _id: '1',
  name: 'Mock Project',
  regionId: mockRegion._id,
  variants: ['Default']
}

export const mockModification = {
  ...mockModel,
  bidirectional: false,
  description: 'Mock description',
  entries: [],
  feed: mockFeed.id,
  _id: '1234',
  name: 'Test Modification',
  routes: [mockRoute.route_id],
  projectId: mockProject._id,
  segments: [mockSegment],
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [true]
}

export const mockProfileRequest = {
  accessModes: 'WALK',
  bikeSafe: 1,
  bikeSlope: 1,
  bikeSpeed: 4.166666666666667,
  bikeTime: 1,
  bounds: mockRegion.bounds,
  bundleId: mockBundle._id,
  date: '2016-01-16',
  defaultExpanded: true,
  directModes: 'WALK',
  egressModes: 'WALK',
  fromLat: 12,
  fromLon: 34,
  fromTime: 25200,
  maxBikeTime: 20,
  maxRides: 4,
  maxTripDurationMinutes: 60,
  maxWalkTime: 20,
  monteCarloDraws: 200,
  name: 'Profile Request Test',
  projectId: mockProject._id,
  reachabilityThreshold: 0,
  regionId: mockRegion._id,
  toTime: 32400,
  transitModes: 'BUS,TRAM,RAIL,SUBWAY',
  travelTimePercentile: 50,
  variantIndex: 0,
  walkSpeed: 1.3888888888888888,
  workerVersion: 'v4.0.0'
}

export const mockRegionalAnalyses = [
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 0,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1,
    travelTimePercentiles: [50]
  },
  {
    ...mockModel,
    zoom: 9,
    bounds: {},
    complete: true,
    deleted: false,
    destinationPointSetIds: ['Jobs_total'],
    status: 'DONE',
    workerVersion: 'v3.2.0',
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    projectId: '1',
    variant: 1,
    bundleId: '1',
    cutoffMinutes: 60,
    cutoffsMinutes: [60],
    _id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 50,
    travelTimePercentiles: [50]
  }
]

const GRID_SIZE = 300
const GRID_MAX = 2000
export const mockGrids = mockRegionalAnalyses.map((ra) => ({
  analysisId: ra._id,
  percentile: 50,
  cutoff: 60,
  zoom: 9,
  width: GRID_SIZE,
  height: GRID_SIZE,
  north: 20024,
  west: 20015,
  data: Array(GRID_SIZE * GRID_SIZE)
    .fill(0)
    .map(() => Math.floor(seededRandom.random() * GRID_MAX)),
  min: 0,
  max: GRID_MAX,
  contains: () => true
}))

export const mockGrid = mockGrids[0]

export const mockStores = {
  init: {
    analysis: {
      copyRequestSettings: true,
      isochroneFetchStatus: false,
      regional: {
        bounds: mockRegion.bounds
      },
      requestsSettings: [mockProfileRequest],
      resultsSettings: [],
      maxTripDurationMinutes: 60,
      travelTimePercentile: 50
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    opportunityDatasets: {
      active: mockOpportunityDataset._id,
      datasets: mockRegion.opportunityDatasets,
      uploadStatuses: []
    },
    queryString: {
      analysisId: mockRegionalAnalyses[0]._id,
      cutoff: 60,
      percentile: 50,
      modificationId: mockModification._id,
      projectId: mockProject._id,
      regionId: mockRegion._id
    },
    region: {
      aggregationAreas: [],
      bundles: [mockBundle],
      regions: [mockRegion]
    },
    regionalAnalyses: {
      analyses: mockRegionalAnalyses,
      grids: mockGrids
    },
    project: {
      activeModificationId: mockModification._id,
      currentProject: mockProject,
      feeds: [mockFeed],
      modifications: [mockModification],
      projects: [mockProject]
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  modificationId: 'Add Trip Pattern',
  title: "Error while applying the modification entitled 'Add Trip Pattern'.",
  messages: ['Headway is not greater than zero.']
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  depth: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

/**
 * This is a function because responseToSurface is destructive and can lead to
 * test leakage if we don't have a brand-new surface each time.
 */
export const createMockSurfaceData = () => {
  const {width, height, depth, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize =
    9 * 4 + // header
    width * height * depth * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * depth)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = depth

  // phew, header is done
  for (let z = 0, position = 9; z < depth; z++) {
    let prev = 0
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++, position++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[position] = pixelVal - prev // values are delta-coded within each depth
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(
    buffer,
    buffer.byteLength - metadataEncoded.byteLength,
    metadataEncoded.length
  )
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

/**
 * Return the value stored in sample z of pixel x, y of the mock surface
 */
export function getMockSurfaceValue(x, y, z) {
  // This used to simply return z, 2 * z and 3 * z when we had an array of
  // [5, 10, 15 . . . 95]th percentiles. However, when we switched to
  // requesting specific percentiles that stopped working. This translates the
  // position in the access grid into a percentile, then subtracts five and
  // divides by five to get the position in the old array, so that tests
  // written assuming the old structure continue to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return (2 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  if (x < 2 && y >= 2) return (3 * (TRAVEL_TIME_PERCENTILES[z] - 5)) / 5
  return 0
}

const opportunityDatasetGridData = new Int32Array(
  mockSurfaceParameters.width * mockSurfaceParameters.height
)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) {
  opportunityDatasetGridData[i] = 1
}
for (
  let i = 3 * mockSurfaceParameters.width;
  i < opportunityDatasetGridData.length;
  i++
) {
  opportunityDatasetGridData[i] = 2
}

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: opportunityDatasetGridData,
  getValue(x, y) {
    return opportunityDatasetGridData[y * mockSurfaceParameters.width + x]
  }
}

export const releaseVersions = ['v3.0.0', 'v2.2.0', 'v2.0.0']
export const allVersions = [
  'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0'
]
