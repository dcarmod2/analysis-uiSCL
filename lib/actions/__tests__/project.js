import nock from 'nock'

import {makeMockStore, mockProject} from 'lib/utils/mock-data'

import * as project from '../project'

describe('actions > project', () => {
  it('create should work', () => {
    nock('http://localhost')
      .post('/api/project')
      .reply(200, {_id: '1'}, {Content: 'json'})

    const action = project.create({
      bundleId: '1',
      name: 'test',
      regionId: '1'
    })
    const store = makeMockStore()
    return store.dispatch(action).then(() => {
      expect(store.getActions()).toHaveLength(3)
    })
  })

  it('deleteProject should work', () => {
    nock('http://localhost').delete('/api/project/1').reply(200, 'deleted')

    const action = project.deleteProject('1', '1')
    const store = makeMockStore()
    return store.dispatch(action).then(() => {
      expect(store.getActions()).toHaveLength(3)
    })
  })

  it('load should work', () => {
    const scope = nock('http://localhost')
      .get('/api/project/1')
      .reply(200, {_id: '1', bundleId: '1'}, {Content: 'json'})

    const action = project.loadProject('1')
    const store = makeMockStore()

    return store.dispatch(action).then((p) => {
      expect(p._id === '1').toBeTruthy()
      expect(store.getActions()).toHaveLength(3)
      scope.done()
    })
  })

  it('saveToServer should work', () => {
    nock('http://localhost')
      .put('/api/project/1')
      .reply(200, {_id: '1'}, {Content: 'json'})

    const action = project.saveToServer(mockProject)
    const store = makeMockStore()
    return store.dispatch(action).then(() => {
      expect(store.getActions()).toHaveLength(3)
    })
  })

  it('set should work', () => {
    expect(project.set()).toMatchSnapshot()
  })

  it('setAll should work', () => {
    expect(project.setAll()).toMatchSnapshot()
  })
})
