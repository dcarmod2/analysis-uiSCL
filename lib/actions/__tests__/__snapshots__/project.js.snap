// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions > project set should work 1`] = `
Object {
  "type": "set project",
}
`;

exports[`actions > project setAll should work 1`] = `
Object {
  "type": "set projects",
}
`;
